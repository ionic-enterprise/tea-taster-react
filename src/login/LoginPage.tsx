import React, { useEffect, useState } from 'react';
import {
  IonContent,
  IonHeader,
  IonIcon,
  IonItem,
  IonLabel,
  IonPage,
  IonSelect,
  IonSelectOption,
  IonTitle,
  IonToolbar,
  isPlatform,
} from '@ionic/react';
import { Device } from '@ionic-enterprise/identity-vault';
import { useHistory } from 'react-router';
import { arrowRedoOutline, lockOpenOutline, logInOutline } from 'ionicons/icons';
import { useSession } from '../core/auth';
import { UnlockMode, useSessionVault } from '../core/vault';

const LoginPage: React.FC = () => {
  const { checkAuthenticationStatus, login, logout, isAuthenticated, error } = useSession();
  const { canUnlock: canUnlockVault } = useSessionVault();
  const { isLocked } = useSessionVault();
  const history = useHistory();
  const [canUnlock, setCanUnlock] = useState<boolean>(false);
  const [hasDeviceSecurity, setHasDeviceSecurity] = useState<boolean>(false);
  const [mode, setMode] = useState<UnlockMode>('NeverLock');

  useEffect(() => {
    canUnlockVault().then((isUnlockable) => setCanUnlock(isUnlockable && isPlatform('hybrid')));
  }, [canUnlockVault]);

  useEffect(() => {
    Device.isSystemPasscodeSet().then(setHasDeviceSecurity);
  }, []);

  // Add useEffect to call checkAuthenticationStatus
  // if Vault is Secure Storage

  useEffect(() => {
    !isLocked && isAuthenticated && history.replace('/tabs');
  }, [isLocked, isAuthenticated, history]);

  const handleLogin = async () => {
    await login(mode);
  };

  const handleUnlock = async () => {
    // Checking the authentication status calls isAuthenticated.
    // That method will prompt the user to unlock the vault in
    // order to gain access to session information to verify
    // if the user is still authenticated.
    await checkAuthenticationStatus();
  };

  const handleRedoLogin = async () => {
    await logout();
    const isUnlockable = await canUnlockVault();
    setCanUnlock(isUnlockable);
  };

  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonTitle>{canUnlock ? 'Unlock' : 'Login'}</IonTitle>
        </IonToolbar>
      </IonHeader>
      <IonContent className="main-content">
        <IonHeader collapse="condense">
          <IonToolbar>
            <IonTitle size="large">{canUnlock ? 'Unlock' : 'Login'}</IonTitle>
          </IonToolbar>
        </IonHeader>
        {!canUnlock ? (
          <>
            <IonItem>
              <IonLabel>Session Locking</IonLabel>
              <IonSelect onIonChange={(e) => setMode(e.detail.value! as UnlockMode)} value={mode}>
                {hasDeviceSecurity && <IonSelectOption value={'Device'}>Device Security</IonSelectOption>}
                <IonSelectOption value={'SessionPIN'}>Session PIN Unlock</IonSelectOption>
                <IonSelectOption value={'NeverLock'}>Never Lock Session</IonSelectOption>
                <IonSelectOption value={'ForceLogin'}>Force Login</IonSelectOption>
              </IonSelect>
            </IonItem>
            <div className="unlock-app ion-text-center" onClick={() => handleLogin()}>
              <IonIcon icon={logInOutline} />
              Sign In
            </div>
          </>
        ) : (
          <>
            <div className="unlock-app ion-text-center" onClick={() => handleUnlock()}>
              <IonIcon icon={lockOpenOutline} />
              Unlock
            </div>
            <div className="unlock-app ion-text-center" onClick={() => handleRedoLogin()}>
              <IonIcon icon={arrowRedoOutline} />
              Redo Sign In
            </div>
          </>
        )}

        {error && <div className="error-message">{error}</div>}
      </IonContent>
    </IonPage>
  );
};

export default LoginPage;
