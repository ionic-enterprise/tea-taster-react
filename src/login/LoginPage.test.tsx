import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import LoginPage from './LoginPage';

describe('<LoginPage />', () => {
  it('renders consistently', () => {
    const { asFragment } = render(<LoginPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the title', () => {
    render(<LoginPage />);
    const titleElements = screen.getAllByText('Login');
    expect(titleElements).toHaveLength(2);
  });

  describe('error messages', () => {
    it('displays an error if the e-mail address is dirty and empty', async () => {
      // render(<LoginPage />);
      // const email = await waitFor(() => screen.getByLabelText('Email Address'));
      // await waitFor(() => fireEvent.change(email, { target: { value: '' } }));
      // await waitFor(() => fireEvent.blur(email));
      // console.log(email);
      //expect(email).toHaveTextContent(/Email address is a required field/);
    });

    it('displays an error if the email address has an invalid format', async () => {});

    it('displays an error message if the password is dirty and empty', async () => {});
  });

  describe('sign in button', () => {
    it('starts disabled', () => {});

    it('is disabled with just an email address', () => {});

    it('is disabled with just a password', () => {});

    it('is enabled with both an email address and password', async () => {
      // render(<LoginPage />);
      // const button = (await waitFor(() => screen.getByText('Sign In'))) as HTMLIonButtonElement;
      // const email = await waitFor(() => screen.getByLabelText('Email Address'));
      // const password = await waitFor(() => screen.getByLabelText('Password'));
      // await waitFor(() => fireEvent.change(password, { target: { value: 'test@ionic.io' } }));
      // await waitFor(() => fireEvent.change(email, { target: { value: 'test@ionic.io' } }));
      // expect(button.disabled).toBeFalsy();
    });
  });
});
