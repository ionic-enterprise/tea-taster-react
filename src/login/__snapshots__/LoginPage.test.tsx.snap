// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> renders consistently 1`] = `
<DocumentFragment>
  <div
    class="ion-page"
  >
    <ion-header
      class="md header-md header-collapse-none"
      role="banner"
    >
      <!---->
      <ion-toolbar
        class="toolbar-title-default md in-toolbar"
      >
        <ion-title
          class="md title-default"
        >
          Login
        </ion-title>
      </ion-toolbar>
    </ion-header>
    <ion-content
      class="main-content md"
      style="--offset-top: 0px; --offset-bottom: 0px;"
    >
      <ion-header
        class="md header-md header-collapse-condense"
        collapse="condense"
        role="banner"
      >
        <!---->
        <ion-toolbar
          class="toolbar-title-large md in-toolbar"
        >
          <ion-title
            class="md title-large"
            size="large"
          >
            Login
          </ion-title>
        </ion-toolbar>
      </ion-header>
      <form>
        <ion-list
          class="md list-md"
        >
          <ion-item
            class="item-interactive item-input item md item-fill-none in-list ion-focusable item-label item-label-floating"
          >
            <ion-label
              class="sc-ion-label-md-h sc-ion-label-md-s md label-floating label-no-animate"
              id="ion-input-2-lbl"
              position="floating"
            >
              E-Mail Address
            </ion-label>
            <ion-input
              class="sc-ion-input-md-h sc-ion-input-md-s md"
              data-testid="email-input"
              type="email"
            >
              <input
                aria-labelledby="ion-input-2-lbl"
                autocapitalize="off"
                autocomplete="off"
                autocorrect="off"
                class="native-input sc-ion-input-md"
                name="ion-input-2"
                placeholder=""
                type="email"
              />
            </ion-input>
          </ion-item>
          <ion-item
            class="item-interactive item-input item md item-fill-none in-list ion-focusable item-label item-label-floating"
          >
            <ion-label
              class="sc-ion-label-md-h sc-ion-label-md-s md label-floating label-no-animate"
              id="ion-input-3-lbl"
              position="floating"
            >
              Password
            </ion-label>
            <ion-input
              class="sc-ion-input-md-h sc-ion-input-md-s md"
              data-testid="password-input"
              type="password"
            >
              <input
                aria-labelledby="ion-input-3-lbl"
                autocapitalize="off"
                autocomplete="off"
                autocorrect="off"
                class="native-input sc-ion-input-md"
                name="ion-input-3"
                placeholder=""
                type="password"
              />
            </ion-input>
          </ion-item>
        </ion-list>
        <div
          class="error-message"
          data-testid="errors"
        >
          <div />
          <div />
          <div />
        </div>
      </form>
    </ion-content>
    <ion-footer
      class="md footer-md"
      role="contentinfo"
    >
      <!---->
      <ion-toolbar
        class="ion-color ion-color-secondary md in-toolbar"
        color="secondary"
      >
        <ion-button
          aria-disabled="true"
          class="md button button-full button-solid in-toolbar in-toolbar-color button-disabled ion-activatable ion-focusable"
          data-testid="submit-button"
          disabled=""
          expand="full"
        >
          Sign In
          <ion-icon
            class="md"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Log In</title><path d='M192 176v-40a40 40 0 0140-40h160a40 40 0 0140 40v240a40 40 0 01-40 40H240c-22.09 0-48-17.91-48-40v-40' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/><path stroke-linecap='round' stroke-linejoin='round' d='M288 336l80-80-80-80M80 256h272' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
            role="img"
            slot="end"
          />
        </ion-button>
      </ion-toolbar>
    </ion-footer>
  </div>
</DocumentFragment>
`;
